sklearn/preprocessing/tests/test_discretization.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import absolute_import
+
+import numpy as np
+import warnings
+
+from sklearn.preprocessing import KBinsDiscretizer
+from sklearn.utils.testing import (
+    assert_array_equal,
+    assert_raises,
+    assert_warns
+)
+
+X = np.array([[-2, 1, -4, -1],
+              [-1, 2, -3, -0.5],
+              [0, 3, -2, 0.5],
+              [1, 4, -1, 2]])
+
+
+def test_fit_transform():
+    dis = KBinsDiscretizer(n_bins=3).fit(X)
+    expected = [[0, 0, 0, 0],
+                [1, 1, 1, 0],
+                [2, 2, 2, 1],
+                [2, 2, 2, 2]]
+    assert_array_equal(expected, dis.transform(X))
+
+
+def test_invalid_n_bins():
+    dis = KBinsDiscretizer(n_bins=1)
+    assert_raises(ValueError, dis.fit_transform, X)
```

Comment:
Because Python's exceptions are so broad in type, `assert_raise_message` or `assert_raises_regexp` often provides substantially better assurances that you've caught the right one!